### ISSUE TITLE
Non-fencing, asynchrony-aware, tools interface

### ISSUE THREAD

**Issue by DavidPoliakoff** [kokkos/kokkos#3723]
This is an issue to request people's thoughts for a version 2 of the Tools interface, one which doesn't require fences, and so doesn't perturb the application beyond all reason. I'm curious about everybody's thoughts, but really want feedback from

- @jrmadsen 
- @daboehme
- @wrwilliams
- @nmhamster 
- @khuck (and the rest of my UO compatriots. I don't know Sameer's user ID, but I hope we can get his feedback)

Update: if you're coming to this topic, [start here](https://github.com/kokkos/kokkos/issues/3723#issuecomment-760243240), the rest of this is a bit of a terrible intro to the problem

As they're the people I know who have dove headfirst into such concepts already. Some hard requirements:

- Tools targeting V1 should work for the foreseeable future
- This interface shall-must-will-shalt be C
- Need to be able to prove that this works with Kokkos Graphs, and with code that uses separate Cuda (for example) instances

Initial thoughts: one question I'm relatively confident in an answer to, but wanted to raise, is whether it's Kokkos or the tool that's responsible for managing things like the CUDA event API. That feels like our problem, asking every tool to write the same code has never been a favorite of mine.

But I think the answer to that question shapes the most important parts of the design, if we decide it's the Tool's problem to look at the underlying asynchrony, it's a hugely different design.

In either branch of this tree, I think what winds up happening is that v2 callbacks also return a decision on whether to fence. Pseudocode for a design here would look like

```c++

struct toolResponse {
  int should_fence; // int-as-bool, can use bool if people prefer
  int unused_1; // allow for us to expand what a tool can tell us later without changing interface
 ...
 int unused_n;
}

struct EventSet { // currently exists, showing some changes
  beginParallelForFunction begin_v1;
  v2BeginParallelForFunction begin_v2; // the v2 function takes a toolResponse* as an arg, otherwise is as in v1
  ...
}

void Profiling::initialize() {
  events.begin_v2 = dlsym("kokkosp_begin_parallel_for_v2");
  if(begin_v2 == nullptr) {
    events.begin_v1 = dlsym("kokkosp_begin_parallel_for");
  }
}

toolResponse create_default_tool_response(){
  return toolResponse { 1 }; // should_fence = true
}

void Profiling::beginParallelFor(all_the_args_begin_parallel_for_takes) {
 if((begin_v1==nullptr) && (begin_v2==nullptr)) { return; }
 auto toolResponse = create_default_tool_response();
  if(begin_v2 != nullptr) {
    (*events.begin_v2)(all_the_v1_args, &toolResponse);
  }
  else (begin_v1!=nullptr){
      (*events.begin_v2)(all_the_v1_args);
  }
  if(toolResponse.shouldFence != 0) {
    Kokkos::fence();
  }
}
```

I could see slight tweaks there, perhaps a tool doesn't return _whether_ to fence, but _what_ ("you need to fence the GPU").

Also, I've basically left this huge gap, where some magical solution figures out how to hide the CUDA event API, the HIP equivalent, OpenMPTarget's stuff, whatever SYCL thinks up... I'm going to be honest and say I don't have a good idea there. That's a place I'd love to lean on the expertise of the community. I have a feeling we have an enum-union around a set of different [somethings], and the tool looks at whether it can handle the kind of asynchrony, based on the enum, to decide whether to fence. I don't know though, my best answer so far is to tag smart people and see what they come up with.

So the asks in this issue are

- Am I missing any requirements? Are any of my requirements wrong?
- Does the toolResponse part of the interface sound reasonable?
- How on earth are we going to abstract these different asynchronous interfaces?

**Comment by khuck** [comment]
@DavidPoliakoff A few questions for clarification before I address anything else...

When you say "CUDA event API" - do you mean:
 - The host-side runtime API callbacks
 - The host-side driver API callbacks
 - The device-side asynchronous activity

...regardless, we already support all of them.  If you are intercepting some/all of them, that blinds the tools...

**Comment by daboehme** [comment]
"CUDA event API" could also be the events one can insert into CUDA streams. Tools have little control over that.

**Comment by khuck** [comment]
not to mention there are several undocumented "conflicts" when enabling different types of CUPTI activity support.  For example, you can either see memory activity or environment activity, but not both.

**Comment by daboehme** [comment]
I'm also lacking a bit of context here. The API is supposed to be "asynchronous" with respect to what? From the tools perspective, the callbacks (at least right now) are synchronous - they record a timestamp or whatever exactly when they are called. How is the tool supposed to know if Kokkos needs a fence or not?

**Comment by nmhamster** [comment]
Few comments - the CUDA API is only one thing we need to worry about, we have other vendors and OMPT, custom tools and all sorts of things we are trying to map into. But, I think your questions and approach is really topical (thank you for bringing it up). I think there is merit in Kokkos querying the tool on successful load to ask the tool which options it wants the runtime to preserve (for instance, do you want fencing per kernel or per region or per..?). We could over time add more and more to that, I like the idea of a struct but perhaps this could be a list of properties and values (e.g. KOKKOS_PERFORM_FENCE = {0, 1}) so we can add additional properties in a long list and not require versioning the struct (we can talk more offline)? The C-basis is important because this is how many vendor tools want to interact.

For the discussion about reproduced code per tool, one option is providing a Kokkos-Tools-Cuda overlay which contains commonly implemented features that CUDA-specific tools would inherit from, we could do the same for HIP, for OpenMP, for etc etc? This could remove duplicated code and reduce development time, plus you could put nice Kokkos idioms in there which we would support.

Just to re-iterate, we have lots of solutions for any specific hardware type (e.g. GPUs) the interface in Kokkos is intentionally not aligned to any specific vendor, programming model or hardware options. I personally want to see that continue (including being agnostic to the tools it connects to).

**Comment by nmhamster** [comment]
@daboehme - Kokkos needs to know whether to fence, the tool knows what it is connecting to on the outside world and whether this needs a fence right?

**Comment by jrmadsen** [comment]
Why not just have the Kokkos pass a function pointer to the tools so that the tool can call `Kokkos::fence()` if it needs it? 

```cpp
using kokkos_fence_func_t = void (*)();

static kokkos_fence_func_t my_kokkos_fence = nullptr;

void kokkosp_set_fence_function(kokkos_fence_func_t _fence) { my_kokkos_fence = _fence; }

void kokkosp_parallel_end_v2(...)
{
    my_kokkos_fence();  
    // ... do something ...
}
```

This way kokkos isn't littered with checks and complex queries and the tools that need a fence can just add a dlsym function and the tools that don't need fencing can just omit the providing `kokkosp_set_fence_function`.

> For the discussion about reproduced code per tool, one option is providing a Kokkos-Tools-Cuda overlay which contains commonly implemented features that CUDA-specific tools would inherit from, we could do the same for HIP, for OpenMP, for etc etc? This could remove duplicated code and reduce development time, plus you could put nice Kokkos idioms in there which we would support.

**Comment by jrmadsen** [comment]
> For the discussion about reproduced code per tool, one option is providing a Kokkos-Tools-Cuda overlay which contains commonly implemented features that CUDA-specific tools would inherit from, we could do the same for HIP, for OpenMP, for etc etc? This could remove duplicated code and reduce development time, plus you could put nice Kokkos idioms in there which we would support.

@nmhamster this is pretty much the _entire idea_ behind [timemory](https://github.com/NERSC/timemory)... You write a component struct which performs one particular measurement or action, like adds a cuda event to a cuda stream, or writes a log message or records values from CUPTI/PAPI/etc., calls Caliper/TAU, etc. and then you bundle the actions you want together into a tuple-like structure and pass both common info (e.g. a label) and component-specific into it (e.g. a `cudaStream_t`) and the bundle handles the negotiation of how to call each through template meta-programming, e.g. 

```cpp
struct foo
{
    void start() { ... }
    void store(int) { ... }
};

struct bar
{
    void start(cudaStream_t ...) { ... };
};

template <typename...> 
class component_tuple;

cudaStream_t _stream;

component_tuple<foo, bar> _bundle { "label" };  

_bundle.start(_stream);   // <-- calls foo::start() and bar::start(cudaStream_t)
_bundle.store(10);        // <-- calls foo::store(int), don't call anything on bar
```

So if you want something like a roofline, you can combine a wall-clock component with a hardware counter component in either a new component or you can bundle them and store the bundle for processing later, etc.

**Comment by jrmadsen** [comment]
And I can say with certainty, creating a common set of tool meta-classes is well beyond the scope of Kokkos at this point.

**Comment by jrmadsen** [comment]
> I'm also lacking a bit of context here. The API is supposed to be "asynchronous" with respect to what? From the tools perspective, the callbacks (at least right now) are synchronous - they record a timestamp or whatever exactly when they are called. How is the tool supposed to know if Kokkos needs a fence or not?

@DavidPoliakoff correct me if I am wrong but I think the idea here is similar to "what if you wanted to record how long the kernel took to launch" not how long the kernel took to run

**Comment by wrwilliams** [comment]
Let me make sure I'm not missing something: is the entire discussion of "access to the (without loss of generality) CUPTI interface" solely based on the ability to call/not call `Kokkos::fence`, or are we additionally talking about enriching the tools interface with more data from the Kokkos device's own tools interface? Because those are IMO very different things.

If I understand correctly, the goal here is something like the following:

* Someone (Kokkos or tool) registers for the (WLOG) CUPTI Activity API, and gets asynchronously notified thereby that `Kernel X` ran from `Start` to `End`
* The Kokkos Tools Interface is notified (directly or from the tool) that this kernel has completed, and translates that kernel completion to meaningful form
* The Kokkos Tools Interface then delivers the (WLOG) `end_parallel_for` callback (extended with timestamp data because that can be asynchronous now?)
* Unlike the present situation, no per-kernel (or at least per-Kokkos-operation) fence calls are required unless the tool wants to force synchronization (e.g. a tool using the CUPTI10 profiling interface and serializing all kernels as a consequence)

@DavidPoliakoff do I have the sense of things here?

**Comment by DavidPoliakoff** [comment]
@wrwilliams , that's a possible implementation, but I apparently need to back way out and give people a "why this all matters." I'll address individual comments separately, but I confused everybody, so lets start there.

First, the way the current interface works. All of our `begin` events (and push_region) have a _global_ Kokkos::fence(). Every Cuda stream, every memory system, every everything fences. We call this before we call your callbacks. All of our end events (and pop_region) also globally fence.

This is inconvenient for users, it massively blows up the time we spend, with tools enabled, even if those tools don't do anything. But I wouldn't change this just out of inconvenience. Broadly speaking, the team sees Kokkos as having more complex asynchronous behaviors in the future. If you look at the increase in compute speeds on GPU's, and the much tamer decrease in kernel launch latencies, it's pretty clear that efficiently launching asynchronous work is going to be critical for performance.

If we hit such a world, a model where tools globally fence Kokkos completely misses the mark. One of the things we will want tools to be able to tell us is how effective our interleaving of different execution streams actually is. So, to frame the objective directly, *I want to get to a place where the Kokkos tools interface can tell users about highly asynchronous models.* Everything else was me taking a wild stab in the dark in that direction. In standard me fashion, I started about mechanism way too early.

**Comment by DavidPoliakoff** [comment]
@daboehme @khuck , I give a little detail above. But what I'm really chasing is the ability to let you guys time an aynchronous model, _without_ the current global fences in Kokkos we require to make it so that tools just think in terms of starting and stopping "the timer stack."

Both of you know the Cupti interface better than I do (ditto OMPT). Suppose I wanted timing data from a code that used N conceptual "streams." Suppose I'm a fanatic about this remaining a C callback interface (because I am). Suppose the interface can't mention Cupti at the callback level, I'm not going to pass you a `cupti[Anything]_t*` (though perhaps an enum-union, one flavor of which might be cupti-like). What would you want that interface to look like?

**Comment by DavidPoliakoff** [comment]
@nmhamster : completely with you on this interface never mentioning a vendor API. I like the idea of common management functions though, that's likely something we need, in fact Bill and I have been talking about that in a few other contexts

**Comment by DavidPoliakoff** [comment]
> > I'm also lacking a bit of context here. The API is supposed to be "asynchronous" with respect to what? From the tools perspective, the callbacks (at least right now) are synchronous - they record a timestamp or whatever exactly when they are called. How is the tool supposed to know if Kokkos needs a fence or not?
> 
> @DavidPoliakoff correct me if I am wrong but I think the idea here is similar to "what if you wanted to record how long the kernel took to launch" not how long the kernel took to run

Not really, but that would be a helpful side-effect. The kernel launch latency simulations are more about _artificially_ inflating kernel launch latencies, so that we can provide real numbers to vendors that say "if you increase our launch latency by a factor of 2, you induce a 30% slowdown and make Stan sad"

The "passing a function pointer" interface isn't a bad one, that's well worth thinking about

**Comment by wrwilliams** [comment]
So, the big question as I see it (from whose answer many other questions derive): is it a hard requirement that v2 have the same synchronous-callback interface as v1, or is an asynchronous callback API or activity buffer API something we could all live with?

As I see it the options are basically threefold: synchronous callbacks (like v1) with maximal relaxation of fences/responsibility shifted to the tool to manage concurrency, asynchronous callbacks but still one callback per event (adding timestamp/location parameters to the interface, with the option to deliver callbacks from a single dedicated thread), or asynchronous/batch (user provides a buffer and a callback, Kokkos calls the callback when the buffer is full and allows some sort of `GetNextEvent` API).

Synchronous callbacks have the virtue of conceptual simplicity, but they're also going to be the most intrusive--they're overhead per-kernel and at the entry/exit points. Asynchronous callbacks allow you to delay the tools calls until they're convenient and move them onto a separate thread if that's helpful. And an activity buffer allows you to perform expensive operations at the Kokkos<->tool boundary once per buffer flush instead of once per callback. If "Kokkos fences on tools interface calls" is still a good invariant and you want to make it performant, some form of batch delivery is what you need...

You can of course build an activity buffer internally and implement an async callback interface over the top of it; I think that might be my aesthetically preferred solution.

**Comment by khuck** [comment]
OK, I _think_ I understand what you're asking.

From the perspective of both TAU and to a lesser extent APEX (and the libraries they depend on, like OTF2), the fence isn't required at all. That's because in order to maintain the "callstack integrity" of the host/calling CPU thread, any asynchronous activity (whether "fire-and-forget", "launch a kernel on device X, stream/queue Y" or "create/schedule/launch N OpenMP tasks and wait for them all to finish") is measured in a totally different hardware context (different OS thread, different GPU/stream "thread").  We maintain some breadcrumbs to link the create call on context A with the schedule/launch/execution on context B, but they aren't _measured_ in the same physical/logical context.  The analysis tool can potentially merge the contexts together later (like I think HPCToolkit does), but measurement is handled separately.  This approach avoids "serializing" the application and perturbing performance tremendously (and providing mostly meaningless performance results).

Aside: An alternative (say if you are tracing) is to buffer all the events in a window, and sort them by timestamp, _hoping_ that some kernel you launched has finished _and_ been reported by CUPTI to the tool by the time you close/end that launch context.  OTF2 (and most other tracing approaches) would barf as soon as timestamps are reported out of order.  There are several down sides to this approach.

APEX is a different, because it was designed for asynchronous runtimes and it doesn't need to maintain a calling context (because it's somewhat meaningless).  There is no CPU/core/thread/device/stream context per se, just a dependency graph.  ..._unless_ OTF2 tracing is enabled, in which case a calling context _is_ maintained (because the library requires it).  Which means it does segregate the measurements into different physical/logical contexts, like TAU does.

I hope I understood your explanation correctly, and didn't just muddy the waters.

**Comment by DavidPoliakoff** [comment]
@khuck FYI I'm not ignoring this, I'm looking for the time to explore it. Let me try to put it concretely, I'm going to show you what parallel_for kinda looks like

```c++
template<class... LotsOfTrash>
void parallel_for(const std::string& name, LotsOfTrash... trash) {
  int64_t kID;
  launch_begin_parallel_for_event(name.c_str(), devId, &kID); // this currently fences
  launch_the_kernel(trash...);
  launch_end_parallel_for_event(kID); // this currently fences
}
```

What I think you're telling me is that TAU and APEX would be capable of just magically making the asynchronous behavior work, if we'd do you the favor of not fencing so much. You'd handle the CUPTI/rocWhatever/WhateverSyclDevelops, and give us profiles using our string kernel names. Regarding the aside: how similar is this to Bill's activity buffer API? To me they sound similar, which in performance tools culture probably means they're utterly unrelated ;)

@wrwilliams : I'm not at all tied to a synchronous callback mechanism. I need vendor-agnosticism and for any callbacks I dlsym (if any exist) to be C. If a tool doesn't support the new interface, we should be able to drop back to our current one. I need people like yourself and Kevin to, if not agree, be willing to say "close enough to what I need." I need to be able to understand the solution and convince myself it's the right idea. If I get hit by a bus, the solution should be equally comprehensible to my replacement. That's kinda my spec sheet, outside of that we're starting with a fairly free discussion of what an interface should do.

In terms of concepts, the activity buffer idea has some appeal, but I worry that to do it effectively the Kokkos Core side of this might need to get more into the tools business than I'd like. I hope you'll forgive the snark in "if I see a call to CUPTI in the kokkos/kokkos repo something has gone terribly wrong, @nmhamster is going to throw something at me." It might be feasible to write an activity buffer API without such a thing, just passing begin-enqueue end-enqueue timestamps and letting the tool sort stuff out, but I really don't know.

If I have Kevin's thoughts right, there's a huge appeal there in that fallback and implementation are just stupidly simple, in the "David-gets-hit-by-a-bus" branch of the decision tree, a replacement could figure out that additional "should_fence" bool pretty quickly.

I'm leaning towards "beg for some of Kevin's time (🍺 ? 🥃 ?), put together a prototype with TAU and/or APEX, use that to inform design" but can easily be swayed on that

**Comment by jrmadsen** [comment]
> I need vendor-agnosticism and for any callbacks I dlsym (if any exist) to be C 

I think it is important to remember here that all the _functions should have C linkage_, not the entire function signature. I personally think if we want to support async in Kokkos, the natural choice is to handle the asynchronousity via the lifetime of a simple virtual object within Kokkos. Consider:

```cpp
namespace Kokkos
{
namespace Tools
{
struct BaseHandler
{
    BaseHandler() = default;
    virtual ~BaseHandler() = default;
    virtual void fence() = 0; 
};

struct ParallelForV1
{
    ParallelForV1(...) 
    { 
        kokkosp_begin_parallel_for(..., &m_idx); 
    }

    ~ParallelForV1() {}

    virtual void fence() final
    {
        ::Kokkos::fence();
        kokkosp_end_parallel_for(m_idx);
    }

    uint64_t m_idx;
};

struct ParallelForV2
{
    ParallelForV2(...) 
    { 
        kokkosp_begin_parallel_for(..., &m_idx); 
    }

    ~ParallelForV2()
    {
        kokkosp_end_parallel_for(m_idx);
    }

    virtual void fence() final {}

    uint64_t m_idx;
};
}  // namespace Tools

template<class... LotsOfTrash>
void parallel_for(const std::string& name, LotsOfTrash... trash) {
  Tools::BaseHandler* _handle = get_tools_parallel_for_object();
  add_handle_to_ImplParallelForObject(_handle);
  launch_the_kernel(trash...);
}
}
```

All of our kernels are tied to an object with an `operator()` so we can tie the destruction of `Tools::BaseHandle` to the destruction of the function object whose `operator()` executes the kernel. If a tool is developed in pure C, we set `get_tools_parallel_for_object()` to return the object which wraps those traditional calls but when a tool is developed in C++, they could provide something like this:

```cpp
// simply redeclare the base
namespace Kokkos
{
namespace Tools
{
struct BaseHandler
{
    BaseHandler() = default;
    virtual ~BaseHandler() = default;
    virtual void fence() = 0; 
};
}
}

struct MyCustomHandler : Kokkos::Tools::BaseHandle
{
    // ... etc. ...
};

extern "C"
{
Kokkos::Tools::BaseHandler* kokkosp_parallel_for_object(...)
{
    return new MyCustomHandler(...);
}
}
```

Any potential issues with this I am overlooking?

**Comment by DavidPoliakoff** [comment]
@jrmadsen: these are legitimate questions I don't know the answer to for sure.

 Don't we open ourselves up to name mangling, object layout, and other differences if the tools return a BaseHandler? Like, if I build the app with Clang but the tool with GCC, am I guaranteed that the vtable and everything of each object will be compatible?

As well as giving up on the ability of (for example) TAU to not need to include a Kokkos header that defines BaseHandle? I'm really nervous to see C++ classes at the callback level, without a very good reason I'd prefer to avoid that, even to the point of preferring structs containing function pointers

**Comment by jrmadsen** [comment]
> name mangling

No, i don't think so, `Kokkos::Tools::BaseHandler` isn't in the mangled name bc it is C linkage. 

> object layout

I don't think so. The base class has no data.

> am I guaranteed that the vtable and everything of each object will be compatible?

I think so, I think in the example above, creating the object in the tool defines the vtable there -- similar to how you can use dynamic polymorphism on the GPU if you do `new` on the GPU.

> As well as giving up on the ability of (for example) TAU to not need to include a Kokkos header that defines BaseHandle?

Well, the idea here was the you would _only_ inherit from `BaseHandle` which, as long as it fixed -- just like all the other callback signatures -- you don't need a header.

**Comment by DavidPoliakoff** [comment]
Lets drill in on vtables. Kokkos will have a view of how a BaseHandler is laid out. The tool with have a view of how BaseHandlers are laid out. If a BaseHandler goes out of scope within Kokkos Core (or we call a function_, then it will say "ah, I know where the destructor is located!" If that doesn't match up with the tool's idea of layout, when the tool creates one you'll have confusion, no? Or do we never destroy/call anything from within Kokkos Core?

**Comment by DavidPoliakoff** [comment]
I also am worried about this "just include the BaseHandler definition" part, but I'll see how tool developers feel about that. We do this already with SpaceHandle, but I frankly think that's a mistake

**Comment by jrmadsen** [comment]
In this `BaseHandle` case, I think it is a very safe assumption that there is ABI-compatibility. I mean, if this wasn't the case, think about how often you'd have to re-compile entire software stacks. Plus, using `BaseHandle` here is a theoretical feature C++ tools could use -- using `BaseHandle` internally in Kokkos is definitely something I think we should consider because I think it would greatly simplify handling the tools

**Comment by jrmadsen** [comment]
Also, again, just to be clear, "include the BaseHandle definition" would _only_ be done if you wanted to do something like `kokkosp_parallel_for_object(...)`, if they provided `kokkosp_begin_parallel_for` and `kokkosp_end_parallel_for` then internally, you wrap those calls into a derived `BaseHandle` object.

**Comment by jrmadsen** [comment]
Yeah, I just did a bit more "research" on vtable implementations (i.e. googled "virtual table wiki"). Basically, from what I gather, the compiler injects a function pointer array in the base class -- this array can be either the last or the first member of the base class (so not an issue for `BaseHandle` because it will be the only member) -- and an object's virtual method table will contain the addresses of the object's dynamically bound methods. So in other words, when `MyCustomHandler` is allocated, it sets the function pointers in the base class to call its functions. That sounds about as safe as doing it manually in this situation to me.

**Comment by DavidPoliakoff** [comment]
Are there guarantees about order, though? If I have 8 virtual functions, is there any guarantee they appear in the vtable in the same order? And is that specified somewhere in a standard, or are we relying on "everybody does it this way?"

That aside, are we saying that to play in the new regime, you need to be a C++ tool? It feels like that is true, and that would preclude some tools.

**Comment by jrmadsen** [comment]
> Are there guarantees about order, though? If I have 8 virtual functions, is there any guarantee they appear in the vtable in the same order? And is that specified somewhere in a standard, or are we relying on "everybody does it this way?"

I am not sure on how the array is actually implemented. I think realistically there is a 0% chance that it is just dumb 1D array and would bind `void fence()` to the vtable entry for a dtor and vice-versa. 

> That aside, are we saying that to play in the new regime, you need to be a C++ tool?

Not at all, I am saying: imagine we adopted this method internally as a way to simplify having to deal with checks for null function pointers, whether to fence, call v1 vs. v2 vs. v725, etc. Then, _theoretically_, we could add some new dlsym callbacks for tools _which choose to do so_, which implement their tools via the object scoping instead of paired begin/end callbacks.

**Comment by wrwilliams** [comment]
> In terms of concepts, the activity buffer idea has some appeal, but I worry that to do it effectively the Kokkos Core side of this might need to get more into the tools business than I'd like. I hope you'll forgive the snark in "if I see a call to CUPTI in the kokkos/kokkos repo something has gone terribly wrong, @nmhamster is going to throw something at me." It might be feasible to write an activity buffer API without such a thing, just passing begin-enqueue end-enqueue timestamps and letting the tool sort stuff out, but I really don't know.

Let me be clear here: Kokkos implementing an internal activity buffer is entirely orthogonal to CUPTI activity buffers. What I am suggesting is that Kokkos could internally take the same approach: record a series of event/time/location tuples as they occur with lightweight synchronization (ideally something atomics-based) and when that buffer is full, fence and deliver the events to tools. Kokkos can decide that a parallel region is complete in whatever manner is convenient, which presumably will have less than zero to do with any internal use of CUPTI. :)

> Are there guarantees about order, though? If I have 8 virtual functions, is there any guarantee they appear in the vtable in the same order? And is that specified somewhere in a standard, or are we relying on "everybody does it this way?"

For things that *actually follow the Itanium C++ ABI*, the order is well specified: functions appear in declaration order, default copy/move/destructor at the end in that order.

https://itanium-cxx-abi.github.io/cxx-abi/abi.html#vtable

I'm going to need more coffee to have a proper opinion on the rest of the overnight conversation...

**Comment by DavidPoliakoff** [comment]
Another point I'm discussing with @wrwilliams in Slack, there's something I've been skipping over (Kevin will be shocked)

```c++
template<class... LotsOfTrash>
void parallel_for(const std::string& name, LotsOfTrash... trash) {
  int64_t kID;
  launch_begin_parallel_for_event(name.c_str(), devId, &kID); // this currently fences
  launch_the_kernel(trash...);
  launch_end_parallel_for_event(kID); // this currently fences
}
```

Note that in this model, we aren't waiting on the kernel to finish before calling `end_parallel_for`. So, if we don't fence, you're only going to see an *enqueue/dispatch* event before `end_parallel_for` gets called, unless we put in subtantial infrastructure in Kokkos to follow execution queues, infrastructure I'd hope to avoid.

**Comment by wrwilliams** [comment]
@jrmadsen In view of @DavidPoliakoff's comment above: is the goal of the `ParallelForV2` struct and similar that, by virtue of destructor magic, we'd still get `end_parallel_for` on kernel completion without fencing? Would that destructor correspond to the point where the application can safely use the result of the parallel for? If not, what Kokkos/application-level meaning does it have?

If this approach is a workable way to get kernel completion events (in some form) with zero/minimal fencing, I am 100% behind that aspect of it. I don't have a strong opinion about the "C++ for tools" part of the proposal, which as I understand it is orthogonal to whether Kokkos uses object scoping internally to determine when to send these `end_foo` callbacks: a C++ tool could choose to entirely rewrite how it handles `parallel_for`, but a C tool would have default `FooV2` structs passing everything through a more loosely synchronized implementation of the V1 API no matter what, right? I can try to ABI lawyer what you guys come up with but I don't have strong feelings about the interface design there.

From my perspective: so long as `end_foo` is delivered at a meaningful time and on a meaningful thread (ie the same thread that delivered `begin_foo`), Score-P should be perfectly capable of handling concurrent accesses to its data in a lightweight manner and keeping overhead low. If either "meaningful time" or "meaningful thread" is missing, we need some addition to the interface that tells us when/where the event actually occurred (and as @khuck points out, we need events on each thread to be delivered in-order). It's entirely fine if the meaningful time is later than what CUPTI/equivalent would tell us about kernel completion, as long as that meaningful time is a consistent point that users a) understand and b) are interested in.

**Comment by mhoemmen** [comment]
> If I have 8 virtual functions, is there any guarantee they appear in the vtable in the same order?

I'm pretty sure the only classes that have layout guarantees are "standard-layout classes," which don't have virtual methods or non-static members with virtual methods.  @nliber or @dhollman might like to say more.

**Comment by DavidPoliakoff** [comment]
@mhoemmen we discuss this above. I don't think the C++ standard even guarantees the existence of a vtable. What Jonathan is leaning on is the Itanium ABI, which most/all compilers mostly follow.

**Comment by mhoemmen** [comment]
@DavidPoliakoff Sorry I missed that point.  Btw, I very much look forward to all the improvements in the tools interface!  It's great stuff! : - D

**Comment by daboehme** [comment]
Going back a bit, I agree with @khuck - tools should work fine without the fences.

Generally, I think it's best to simply define the semantics of the callbacks and let the tools / users deal with it. If the parallel region callbacks are asynchronous, great, the parallel region time is your kernel launch time then or whatever the backend did. For CUDA specifically we can also associate CUDA kernels with those region names thanks to CUPTI correlation records. I'm reasonably confident that other APIs like OMPT and rocM have something similar.

The one exception I can think of where you _do_ need fences is the case where you want to record UVM events and associate them with the parallel region. But that's a rather specific use case, and you could work around it by adding another region with an explicit fence around the parallel_for. If you do want to retain the current behavior you could add a config variable like `KOKKOS_PROFILING_SHOULD_FENCE` or `KOKKOS_SERIALIZE` or something and let the users set that if needed.

**Comment by jrmadsen** [comment]
So here are some updates. 

@DavidPoliakoff has sufficiently convinced me that the vtables in tools idea is risky but it is still possible there may be a supplemental interface where you can stick something into a `void*` and get it returned to you in lieu of indexes. 

Also, w.r.t. the internal usage of objects avoid a fence but still making the "end callback" happen asyncly, it may work for the CPU-based parallelism models but for CUDA specifically I was mislead by my initial scan of `Kokkos::Impl::ParallelFor` for CUDA, which does indeed use an `operator()` to execute the kernel (which requires a `this`), which lead me to assume there might be a destructor call happening after the kernel was executed but it appears CUDA does something like malloc + memcpy + free when copying the object over to the GPU, thus bypassing all calls to the ctor and dtor. But using something like `cudaStreamAddCallback` might be the solution here if we can find consistent ways across all Kokkos backends to exploit either using the existence of the `this` pointer in order to call `operator()` to execute the callback when the kernel has completed.

> If you do want to retain the current behavior you could add a config variable like KOKKOS_PROFILING_SHOULD_FENCE or KOKKOS_SERIALIZE or something and let the users set that if needed.

@daboehme Assuming you mean making this a build option, I don't think we want to fix fencing behavior in the Kokkos build because then you have to rebuild kokkos and and the app using kokkos to change the behavior vs. just setting a different profiling library or configuring something in the tool.

**Comment by daboehme** [comment]
> @daboehme Assuming you mean making this a build option, I don't think we want to fix fencing behavior in the Kokkos build because then you have to rebuild kokkos and and the app using kokkos to change the behavior vs. just setting a different profiling library or configuring something in the tool.

I was actually thinking about a runtime option with environment variables or something.

**Comment by wrwilliams** [comment]
> Also, w.r.t. the internal usage of objects avoid a fence but still making the "end callback" happen asyncly, it may work for the CPU-based parallelism models but for CUDA specifically I was mislead by my initial scan of Kokkos::Impl::ParallelFor for CUDA, which does indeed use an operator() to execute the kernel (which requires a this), which lead me to assume there might be a destructor call happening after the kernel was executed but it appears CUDA does something like malloc + memcpy + free when copying the object over to the GPU, thus bypassing all calls to the ctor and dtor. But using something like cudaStreamAddCallback might be the solution here if we can find consistent ways across all Kokkos backends to exploit either using the existence of the this pointer in order to call operator() to execute the callback when the kernel has completed.

HIP mimics CUDA here, so if `cudaStreamAddCallback` is an acceptable level of performance hit, that should cover those two backends. OMPT should likewise give callbacks at the end of parallel regions whether they're on host or device, though I don't know in practice how well OMPT handles multiple tools and @DavidPoliakoff doubtless would prefer not to write OMPT code inside Kokkos. There's got to be a non-tools-interface equivalent though...

**Comment by khuck** [comment]
@wrwilliams I am familiar with this technique of adding `cudaStreamAddCallback` to the command stream, but what does it buy you as a tool?  You know when the kernel finishes, sure.  But you then _still_ have to force CUPTI to flush the buffer.  And you still have to process the async activities that are in the buffer, searching for the one that is associated with your call(s) (based on correlation ID).  You are just now doing it all synchronously, occupying/perturbing the CPU thread.

The simplest solution is for Kokkos to simply synchronize the stream at the end of the parallel region (which I assume it does anyway, to make sure the results are ready for the next block of executable code).  Once the parallel region exits (for which there is a callback), the tool can safely assume that the asynchronous activity is also done.  And can choose to process the buffer at that point.  If the tool doesn't want to assume that Kokkos forced a stream synchronize, it can do it explicitly.

I still say let the tools handle the asynchronous GPU activity, they are doing it already anyway...only one library is going to be allowed to register for callbacks and process buffer activity, and if Kokkos intercepts it all, that prevents the tools from doing any specialized measurement/analysis that they are coded to do.  And that the tool users are expecting.

In short, there's no default requirement for the fence.  If the tool needs to, it'll synchronize the stream/queue.  Make it runtime configurable with an environment variable.

**Comment by khuck** [comment]
@DavidPoliakoff not to belabor the point...but I just ran a few Kokkos unit tests (OpenMP & CUDA) with APEX, and there is a LOT of (what I think is) superfluous synchronization going on.  Every stream synchronization seems to be followed by a device synchronization. or vice versa, I can't really tell...maybe that's a side effect (or actual behavior) of the fencing?  Case in point from zooming in on a Vampir trace of `containers/performance_tests/KokkosContainers_PerformanceTest_Cuda`:
```
Kokkos reduce begin (CPU)
  cuda launch kernel begin/end (CPU & GPU)
  cuda stream synchronize begin (CPU & GPU)
    kernel executes (GPU)
  cuda stream synchronize end (CPU & GPU)
  cuda device synchronize begin/end (CPU & GPU)
kokkos reduce end (CPU)
cuda stream synchronize begin/end (CPU & GPU)
cuda device synchronize begin/end (CPU & GPU)
(repeat, ad nauseam)
```
I know it's a unit test, but if all that sync is for the benefit of the tool (fencing), it's not necessary and kills both performance and meaningful measurement.

**Comment by jrmadsen** [comment]
> I am familiar with this technique of adding cudaStreamAddCallback to the command stream, but what does it buy you as a tool?
> (and other similar comments along this theme)

@khuck One of the things to remember here is that it is not only TAU, APEX, Score-P, Caliper, etc. using the tools interface. It is easy for us to forget bc we work so hard on making these tools with all these cool features but unfortunately lots of people just use the relatively simple tools provided by kokkos and sometimes write their own. And in those cases, it would enormously complicate writing those tools if they had to account for all this different backends in order to make sure the values are relevant. Technically, you could avoid the fence in v1 by just not defining `kokkosp_end_parallel_{for,reduce,scan}`.

> The simplest solution is for Kokkos to simply synchronize the stream at the end of the parallel region (which I assume it does anyway, to make sure the results are ready for the next block of executable code).

Kokkos typically won't sync at all (even on the stream) unless a transfer back to the CPU is required, which in many cases, it isn't.

**Comment by wrwilliams** [comment]
> > I am familiar with this technique of adding cudaStreamAddCallback to the command stream, but what does it buy you as a tool?
(and other similar comments along this theme)

> @khuck One of the things to remember here is that it is not only TAU, APEX, Score-P, Caliper, etc. using the tools interface. It is easy for us to forget bc we work so hard on making these tools with all these cool features but unfortunately lots of people just use the relatively simple tools provided by kokkos and sometimes write their own. And in those cases, it would enormously complicate writing those tools if they had to account for all this different backends in order to make sure the values are relevant.

From a  Score-P perspective, if we really wanted to, we could work quite happily (and I could make the argument it would be better for us) with a Kokkos tools interface that is solely designed to ensure that we can relabel information from the tools interface of the execution space with an appropriate Kokkos descriptor (to be concrete in the CUDA case, a mapping from kernels to their associated parallel_X launches, with whatever level of descriptive detail is necessary and proper). But as @jrmadsen points out, we're not the only or even the primary consumer of the tools interface and it's important that simple one-off tools are able to get end events that are in some way meaningful--without needing to write their own CUPTI code.

As I understand it, registering a stream callback rather than using stream/device synchronize calls allows Kokkos to use more flexible async launch models under the hood and still be aware of when a kernel completes. Basically if you're using stream synchronize to determine completion you're stuck with a model where `parallel_for` spawns a task and waits, whereas if you use the stream callback, you can have an asynchronous launch on a thread that also does CPU computation. My understanding here may well be wrong though.

Now, I do need to think again about the implications this has for us, because in writing this down I realize that our mapping of Kokkos events into OTF2 isn't quite right when kernel execution becomes asynchronous unless the task-per-launch model is what you guys eventually go with...

**Comment by DavidPoliakoff** [comment]
So, I've tried to condense all this to an implementation, based on some conversations with @khuck . At the end of our conversation, we needed two things

1) Kokkos apps need to see their provided kernel names to the tool.
2) Tools need to be able to tell Kokkos not to fence so much

Remember that the code for `parallel_for` looks like

```
void parallel_for(stuff){
  should_i_fence_before_begin? // currently always yes
  if yes
    fence
  send_begin_event
  launch_kernel
  should_i_fence_before_end? // currently always yes
  if yes
    fence
  send_end_event
}
```

Critically, Kokkos _is not aware of when kernels complete_. I think a lot of above discussion thinks that Kokkos knows when a kernel completes, and can just send an end event then. But all Kokkos does is sync an execution space instance (think CUDA stream) when you tell it to.

So, what a tool might do is make both of those "should I fence" questions respond "no." Then all the begin/end bracket will contain is the kernel _submission/enqueuing_. But the tool knows about CUPTI, and can watch for the kernel to end by itself, and then associate the timing with the kernel, without any fencing on the Kokkos side.

I'm sketching out an implementation here (#3799). How this works is that tools can have a "response" to the callback they implement (for brevity, I just did init and begin/end_parallel_for/scan/reduce). Right now it's a struct with one bool, just a "should_fence" guy. There are three levels of specificity for these

1) Kokkos defaults
2) Tool defaults
3) Tool responses

Kokkos defaults to saying "all events should fence." Then, it passes one of these guys to the tool on init. If the tool overrides that ToolResponse, it becomes the default for the rest of the run. Then, at every event, a ToolResponse is created with that new default, and the tool can adjust its behavior.

Critically, for a tool that just wants to say "never fence", it can maintain all of its v1 callbacks for kernel events, and just on init override that ToolResponse.

In C++,

```c++
  template <typename V2Function, typename V1Function, typename... Args>
  Kokkos::Tools::Experimental::ToolResponse call_kokkos_callback(
      const V2Function v2, const V1Function v1, Args... args) {
    Kokkos::Tools::Experimental::ToolResponse response = default_response;
    if (v2 != nullptr) {
      (*v2)(args..., &response);
    } else if (v1 != nullptr) {
      if (response.should_fence) {
        Kokkos::fence();
      }
      (*v1)(args...);
      response.should_fence = false;  // no post-callback fence in the old scheme
    }
    return response;
  }
```

Note that the fencing for V2 happens outside of this function, at the moment, but it does happen

```c++
    auto response = Kokkos::Tools::Experimental::call_kokkos_callback(
        Kokkos::Tools::Experimental::current_callbacks.begin_parallel_for_v2,
        Kokkos::Tools::Experimental::current_callbacks.begin_parallel_for,
        kernelPrefix.c_str(), devID, kernelID);
    if (response.should_fence) {
      Kokkos::fence();
    }
```

This is getting wordy, I'll stop it here, but that's just a sketch of where I could see this heading

**Comment by DavidPoliakoff** [comment]
Have an alternate implementation up here: #3812

Got to be honest, I prefer that implementation. Slightly more weight on the tool, but tool developers are smarter than we are anyway, right?
